libraries/Microsoft.Bot.Builder.Azure/AzureBlobTranscriptStore.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using Microsoft.Bot.Schema;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Azure
+{
+    /// <summary>
+    /// The blob transcript store stores transcripts in an Azure Blob container.
+    /// </summary>
+    /// <remarks>
+    /// Each activity is stored as json blob in structure of
+    /// container/{channelId]/{conversationId}/{Timestamp.ticks}-{activity.id}.json 
+    /// </remarks>
+    public class AzureBlobTranscriptStore : ITranscriptStore
+    {
+        private readonly static JsonSerializer jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings()
+        {
+            NullValueHandling = NullValueHandling.Ignore,
+            Formatting = Formatting.Indented
+        });
+
+        private static HashSet<string> _checkedContainers = new HashSet<string>();
+
+        private Lazy<CloudBlobContainer> Container { get; set; }
+
+        /// <summary>
+        /// Creates an instance of AzureBlobTranscriptStore
+        /// </summary>
+        /// <param name="dataConnectionString">Connection string to connect to Azure Blob Storage</param>
+        /// <param name="containerName">Name of the continer where transcript blobs will be stored</param>
+        public AzureBlobTranscriptStore(string dataConnectionString, string containerName)
+            : this(CloudStorageAccount.Parse(dataConnectionString), containerName)
+        {
+
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="storageAccount">Azure Storage Account to store transcripts</param>
+        /// <param name="containerName">Name of the continer where transcript blobs will be stored</param>
+        public AzureBlobTranscriptStore(CloudStorageAccount storageAccount, string containerName)
+        {
+            if (storageAccount == null)
+                throw new ArgumentNullException(nameof(storageAccount));
+
+            if (String.IsNullOrEmpty(containerName))
+                throw new ArgumentNullException(nameof(containerName));
+
+            this.Container = new Lazy<CloudBlobContainer>(() =>
+            {
+                containerName = containerName.ToLower();
+                var blobClient = storageAccount.CreateCloudBlobClient();
+                NameValidator.ValidateContainerName(containerName);
+                var container = blobClient.GetContainerReference(containerName);
+                if (!_checkedContainers.Contains(containerName))
```

Comment:
`HashSet<T>` is _not_ a threadsafe type and this logic could theoretically be accessed from multiple, concurrent threads. At bare minimum add a `lock(_checkedContainers)` before this logic. Consider possibly switching to `ConcurrentDictionary` and even hanging onto the `CloudBlobContainer` instances in there as the values since they're safe to use from concurrent threads anyway as long as you are not mutating the container itself.